### DISCUSSION TITLE
static-lib example executable print blank result compiled by cl

### DISCUSSION THREAD

**Discussion by LukyGuyLucky** [c3lang/c3c#1726] [General]
With [https://github.com/c3lang/c3c/tree/master/resources/examples/staticlib-test](url) ,the executable of test.c3 print proper result:
```
In adder
Print foo: Foo[1]
5
```
while the executable test.exe of test.c compiled by VS Studio 2022 cl just print blank. 

I don't know what I did wrong.

```
c3c static-lib add.c3
c3c compile-run test.c3 -l add.lib
cl test.c add.lib //under x64 Native Tools Command Prompt
```

Win10 64 bit ,MS Visual Studio 2022,the most updated c3c (as of today).

**Comment by lerno** [comment]
Does the dynlib test work?

**Reply by LukyGuyLucky** [reply]
Yes,just tried.both c3 and c works.

**Reply by lerno** [reply]
Static libraries is the tricky part. Can you see if you add an @init method in the static library if that one is called when you use C?

Like
```c
import libc;
fn void startuptest() @init
{
  libc::puts("Hello from startuptest");
}
```

**Reply by LukyGuyLucky** [reply]
No,it is the same in c; In c3 test ```Hello from startuptest``` printed as well.

**Reply by lerno** [reply]
What if you change this:
```c
int main()
{
	printf("%d\n", adder(1, 4));
	return 0;
}
```
To
```c
int main()
{
  puts("Before");
  printf("%d\n", adder(1, 4));
  puts("After");
  return 0;
}
```
What is printed?

**Comment by LukyGuyLucky** [comment]
Things are weird.
The first thing I did this morning is just to recompile test.c according to your most recent tips ---add ```puts("Before") and puts("After")``` and then re-run the example,so I got ```Before``` printed.Yes,only```Before```.

After that I deleted all generated files and I tried to rebuilt the add.lib from scratch.Now when I tried to rebuild test.c I got a compiler complaints which means Unresolved external symbols SymInitialize,SymCleanup ,etc . 

for test.c3,everything is ok.
```
e:\Learning\c3pack\compiler\c3c-master\c3c-master\resources\examples\staticlib-test>cl test.c add.lib
用于 x64 的 Microsoft (R) C/C++ 优化编译器 19.42.34435 版
版权所有(C) Microsoft Corporation。保留所有权利。

test.c
Microsoft (R) Incremental Linker Version 14.42.34435.0
Copyright (C) Microsoft Corporation.  All rights reserved.

/out:test.exe
test.obj
add.lib
  正在创建库 test.lib 和对象 test.exp
add.lib(add.obj) : error LNK2019: 无法解析的外部符号 SymInitialize，函数 std.os.win32.symbolize_backtrace 中引用了该符号
add.lib(add.obj) : error LNK2019: 无法解析的外部符号 SymCleanup，函数 std.os.win32.symbolize_backtrace 中引用了该符号
add.lib(add.obj) : error LNK2019: 无法解析的外部符号 SymFromAddr，函数 std.os.win32.resolve_backtrace 中引用了该符号
add.lib(add.obj) : error LNK2019: 无法解析的外部符号 SymGetModuleInfo64，函数 std.os.win32.resolve_backtrace 中引用了该 符号
add.lib(add.obj) : error LNK2019: 无法解析的外部符号 UnDecorateSymbolName，函数 std.os.win32.resolve_backtrace 中引用了 该符号
add.lib(add.obj) : error LNK2019: 无法解析的外部符号 SymGetLineFromAddr64，函数 std.os.win32.resolve_backtrace 中引用了 该符号
test.exe : fatal error LNK1120: 6 个无法解析的外部命令
```

**Reply by lerno** [reply]
Maybe you need to manually link to some of those missing libraries. For the above you would need to link with Dbghelp.lib

**Reply by LukyGuyLucky** [reply]
with dbghelp.lib now test.c works as well.

**Reply by lerno** [reply]
So it prints before / after and the part in between?

**Reply by LukyGuyLucky** [reply]
Yes,all printed as expected.
```
Add module initializes here.
Before
In adder
Print foo: Foo[1]
5
After
Add module finalizes here.
```

**Reply by LukyGuyLucky** [reply]
BTW,dgbhelp.lib is needed here which is not so common as expected.
